File: app/apple/ios/HiAgentIOS/PairingRequest.swift
import Foundation

enum PairingRequestError: LocalizedError {
    case invalidLink
    case missingAddress
    case missingCode

    var errorDescription: String? {
        switch self {
        case .invalidLink:
            return "This is not a Hi Agent pairing link."
        case .missingAddress:
            return "The pairing link does not contain a core address."
        case .missingCode:
            return "The pairing link does not contain a pairing code."
        }
    }
}

struct PairingRequest: Identifiable, Equatable {
    let id = UUID()
    let baseURL: String
    let code: String
    let label: String

    static var manual: PairingRequest {
        PairingRequest(baseURL: "", code: "", label: "")
    }

    init(baseURL: String, code: String, label: String) {
        self.baseURL = baseURL
        self.code = code
        self.label = label
    }

    init(url: URL) throws {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false),
              components.scheme?.lowercased() == "hiagent",
              components.host?.lowercased() == "pair"
        else {
            throw PairingRequestError.invalidLink
        }

        let items = components.queryItems ?? []
        guard let rawBaseURL = Self.singleValue(named: "url", in: items),
              !rawBaseURL.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        else {
            throw PairingRequestError.missingAddress
        }
        guard let rawCode = Self.singleValue(named: "code", in: items),
              !rawCode.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        else {
            throw PairingRequestError.missingCode
        }

        baseURL = try CoreClient.normalizeBaseURL(rawBaseURL).absoluteString
        code = rawCode.trimmingCharacters(in: .whitespacesAndNewlines)
        label = Self.singleValue(named: "label", in: items)?
            .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    }

    private static func singleValue(named name: String, in items: [URLQueryItem]) -> String? {
        let matches = items.filter { $0.name == name }
        guard matches.count == 1 else {
            return nil
        }
        return matches[0].value
    }
}

File: app/android/app/src/main/java/com/xiaoyuanzhu/hiagent/android/core/PairingRequest.kt
package com.xiaoyuanzhu.hiagent.android.core

import android.net.Uri
import java.util.UUID

class PairingLinkException(message: String) : Exception(message)

/**
 * A `hiagent://pair?url=…&code=…` link, from the QR the core's Reach view draws
 * or from a link handed to the app by anything else on the handset.
 *
 * The core builds this in `pairing_app_url()` and knows nothing about which
 * platform will read it, so nothing here is Android-specific.
 */
data class PairingRequest(
    val baseUrl: String,
    val code: String,
    val label: String,
    /** Scanning is the intended path, so the sheet can arrive with the camera up. */
    val opensScanner: Boolean = false,
    val id: String = UUID.randomUUID().toString(),
) {
    companion object {
        fun manual() = PairingRequest("", "", "")

        fun scan() = PairingRequest("", "", "", opensScanner = true)

        @Throws(PairingLinkException::class)
        fun fromUri(uri: Uri): PairingRequest {
            if (!uri.scheme.equals("hiagent", ignoreCase = true) ||
                !uri.host.equals("pair", ignoreCase = true)
            ) {
                throw PairingLinkException("This is not a Hi Agent pairing link.")
            }

            val rawBaseUrl = singleValue(uri, "url")
                ?.takeIf { it.isNotBlank() }
                ?: throw PairingLinkException(
                    "The pairing link does not contain a core address.",
                )
            val rawCode = singleValue(uri, "code")
                ?.takeIf { it.isNotBlank() }
                ?: throw PairingLinkException(
                    "The pairing link does not contain a pairing code.",
                )

            val baseUrl = try {
                CoreClient.normalizeBaseUrl(rawBaseUrl).toString()
            } catch (e: CoreClientException) {
                throw PairingLinkException(e.message ?: "The core address is not usable.")
            }

            return PairingRequest(
                baseUrl = baseUrl,
                code = rawCode.trim(),
                label = singleValue(uri, "label")?.trim().orEmpty(),
            )
        }

        /**
         * A repeated parameter is refused rather than resolved. `?code=a&code=b`
         * has no single right answer, and picking one would be guessing about a
         * link that is already malformed.
         */
        private fun singleValue(uri: Uri, name: String): String? {
            val values = try {
                uri.getQueryParameters(name)
            } catch (_: UnsupportedOperationException) {
                return null
            }
            return values.singleOrNull()
        }
    }
}
